src/app/cli/src/commands/compact_command.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use kamu::domain::compact_service::CompactService;
+use kamu::domain::{
+    DatasetRepository,
+    VerificationMultiListener,
+    VerificationOptions,
+    VerificationService,
+};
+use opendatafabric::{DatasetHandle, DatasetRef};
+
+use crate::{CLIError, Command, CompactionMultiProgress, VerificationMultiProgress};
+
+pub struct CompactCommand {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    verification_svc: Arc<dyn VerificationService>,
+    compact_svc: Arc<dyn CompactService>,
+    dataset_ref: DatasetRef,
+    dataset_dir_path: PathBuf,
+}
+
+impl CompactCommand {
+    pub fn new(
+        dataset_repo: Arc<dyn DatasetRepository>,
+        verification_svc: Arc<dyn VerificationService>,
+        compact_svc: Arc<dyn CompactService>,
+        dataset_ref: DatasetRef,
+        dataset_dir_path: PathBuf,
+    ) -> Self {
+        Self {
+            dataset_repo,
+            verification_svc,
+            compact_svc,
+            dataset_ref,
+            dataset_dir_path,
+        }
+    }
+
+    async fn verify_dataset(&self, dataset_handle: &DatasetHandle) -> Result<(), CLIError> {
+        let progress = VerificationMultiProgress::new();
+        let listener = Arc::new(progress.clone());
+        let draw_thread = std::thread::spawn(move || {
+            progress.draw();
+        });
+
+        let result = self
+            .verification_svc
+            .verify(
+                &dataset_handle.as_local_ref(),
+                (None, None),
+                VerificationOptions::default(),
+                listener.begin_verify(dataset_handle),
+            )
+            .await;
+
+        listener.finish();
+        draw_thread.join().unwrap();
+
+        result.outcome.map_err(CLIError::failure)
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl Command for CompactCommand {
+    async fn run(&mut self) -> Result<(), CLIError> {
+        let dataset_handle = self
+            .dataset_repo
+            .resolve_dataset_ref(&self.dataset_ref)
+            .await
+            .map_err(CLIError::failure)?;
+
+        if let Err(err) = self.verify_dataset(&dataset_handle).await {
+            eprintln!(
+                "{}",
+                console::style("Cannot perform compacting, dataset is invalid".to_string()).red()
+            );
+            return Err(err);
+        }
```

Comment:
Probably, misunderstood this commend https://docs.google.com/document/d/1UCFUQ3V_muPbJuhl8dHFf3aRXlXL3yGbekbDwPlnaUA/edit?disco=AAABILoXam0. I was thinking that if only one block is corrupted, user can reset this block manually. And after run compaction. But if we merge corrupted block to fix it will be really dificult.